Feat: `bls` crate with modular `backend`s

shreyas-londhe opened the pull request:
This PR generalizes the use of backends in the `bls` crate. New backends can be added to this crate and can be used by replacing features where the crate is used.

Resolves #61

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grandinetech/grandine?pullRequest=69) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grandinetech/grandine?pullRequest=69) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grandinetech/grandine?pullRequest=69) it.</sub>

shreyas-londhe commented:
Hi @sauliusgrigaitis I have implemented the necessary traits for the `bls` crate and added the `blst` backend as default. Can you please review my changes and let me know of any changes or suggestions you have? 

I would proceed to add more backends starting with `bls12_381` once you have given a ✅ to my current changes. 

Thank you!

povi commented:
**[Suggestion](https://conventionalcomments.org/)**:

You don't need to prefix all the trait names with `Bls*`.
I guess the reason for adding the prefixes was that without them trait names clash with other `bls` type names.
You could drop the `Bls*` prefix from trait names and import them like this:

```
use bls::{
    traits::{PublicKey as _, SignatureBytes as _},
    PublicKey, SignatureBytes,
};
```

We rarely need imported trait names in code. Most of the time importing traits anonymously is better.

povi commented:
**[Suggestion](https://conventionalcomments.org/)**:

All the old code that was wrapping the `blst` crate types was moved to `blst` backend, and the new traits are just the method signatures and nothing more.

When you add a new `bls` backend, much of the code that wraps some newly added bls crate's types will probably be pretty much copy-pasted from `blst` backend with minor changes.

So my suggestion is to try and move as much as possible code from `blst` backend to the trait methods, so it can be reused with different backends.

sauliusgrigaitis commented:
Thanks @shreyas-londhe for PR, but I agree with @povi. I think it's best if you proceed by adding `bls12_381` crate and then you simply refactor your PR so it doesn't repeat the code.

shreyas-londhe commented:
Thanks for the suggestions @povi. I'll add the `bls12_281` crate and then move the repeated code to the traits.

shreyas-londhe commented:
Hi @povi @sauliusgrigaitis, the `bls12_381` crate doesn't have a `hash_to_curve` implementation OOTB, are you guys aware of any existing ones we can use or shall I implement it from scratch?

shreyas-londhe commented:
Hi @povi I made the changes you requested. Can you please tell me how I can test if my changes didn't break anything? Please let me know of any more suggestions you have :)

sauliusgrigaitis commented:
@shreyas-londhe the very first step would be to fix the build. You can run the build locally by executing the commands from the build script https://github.com/grandinetech/grandine/blob/develop/.github/workflows/ci.yml.

shreyas-londhe commented:
@sauliusgrigaitis I'm getting clippy errors which are not from the code I modified. Can you please check that once? I even tried running the tests locally and they fail at code which I have not modified. Can you please help me out on why the CI might be failing?

povi commented on transition_functions/src/phase0/block_processing.rs:
```diff
@@ -485,15 +486,15 @@ mod spec_tests {
         "consensus-spec-tests/tests/mainnet/phase0/operations/block_header/*/*",
-        "consensus-spec-tests/tests/minimal/phase0/operations/block_header/*/*",
-    }
+        "consensus-spec-tests/tests/minimal/phase0/operations/block_header/*/*"
```
I see a TON of changes that are either a space removed, a comma removed or a semicolon added. Neither of these changes are related to this pull request.

What is more, some of these changes you made break macros, like this one. Now it produces error:

```
error: unexpected end of macro invocation
   --> transition_functions/src/phase0/block_processing.rs:494:80
    |
432 |     macro_rules! processing_tests {
    |     ----------------------------- when calling this macro
...
494 |         "consensus-spec-tests/tests/minimal/phase0/operations/block_header/*/*"
    |                                                                                ^ missing tokens in macro arguments
    |
note: while trying to match `,`
   --> transition_functions/src/phase0/block_processing.rs:438:34
    |
438 |             $minimal_glob:literal,
```

povi replied:
Why did you make these changes?

shreyas-londhe replied:
@povi Looks like these were because of running ‘cargo fmt —all’. Let me remove those.

shreyas-londhe replied:
@povi Thanks for pointing this out, I reverted all the formatting changes which are not related to this PR. How do I invoke the CI workflow on my latest commit?

sauliusgrigaitis replied:
@shreyas-londhe you should be able to fork the repo and push the changes to branches in your local repo. This should trigger CI.

sauliusgrigaitis commented:
> @sauliusgrigaitis I'm getting clippy errors which are not from the code I modified. Can you please check that once? I even tried running the tests locally and they fail at code which I have not modified. Can you please help me out on why the CI might be failing?

The build on `develop` branch passes https://github.com/grandinetech/grandine/commits/develop/ . If you don't know which of your changes crashed the build -  the easiest way for you would be to apply your changes incrementally and see which change crashed the build.

shreyas-londhe commented:
Hi @povi @sauliusgrigaitis I fixed all the clippy issues and made sure that the CI passes. Can you please review my PR and let me know of any suggestions?

sauliusgrigaitis commented:
We want to make this inline with https://github.com/grandinetech/rust-kzg naming. So with the same switch both our `bls` and `rust-kzg` will be able to switch to the particular backend. So the name `zkcrypto` needs to be used instead of `bls12_381`. The same is with the other backends such as `arkworks`.

Can you also squash to a single commit after the naming is changed?

sauliusgrigaitis commented:
I also suggest to to enable `zkcrypto` backend when running tests in Gitlab CI either on Macos or on Windows target.

sauliusgrigaitis commented:
> I also suggest to to enable `zkcrypto` backend when running tests in Gitlab CI either on Macos or on Windows target.

I just checked and it doesnt compile with `--features bls12_381`. So it really makes sense to add to Gitlab CI. Simply add it [here](https://github.com/grandinetech/grandine/blob/develop/.github/workflows/ci.yml#L37).

shreyas-londhe commented:
@sauliusgrigaitis The code compile when `default` features are off. I have kept `bls-blst` as the default backend and other backends can be compiled by running `--no-default-features --features <bls-backend>`. Please let me know if this is desired or I should change it have no default feature.

sauliusgrigaitis commented:
@shreyas-londhe  you must test your code before you ask for review. It fails https://github.com/grandinetech/grandine/commit/b3577bc0e4fb32f19d5cd008edf87578daf654f1

shreyas-londhe commented:
Sorry for the confusion @sauliusgrigaitis, as you mentioned I changed the feature names to match `rust-kzg`.

```bash
# in `grandine` root
cd bls

# builds with `blst` backend
cargo build 

# builds with `zkcrypto (bls12_281)` backend
cargo build --no-default-features --features bls-zkcrypto
```

sauliusgrigaitis commented:
> Sorry for the confusion @sauliusgrigaitis, as you requested I changed the feature names to match `rust-kzg`.
> 
> ```shell
> # in `grandine` root
> cd bls
> 
> # builds with `blst` backend
> cargo build 
> 
> # builds with `zkcrypto (bls12_281)` backend
> cargo build --no-default-features --features bls-zkcrypto
> ```

The switch needs to work on the top level, not in `bls` crate. You need to edit `.github/workflows/ci.yml` and add something like this https://github.com/shreyas-londhe/grandine/compare/feat/modular-bls...grandinetech%3Agrandine%3Afeat/modular-bls . Then push this change to your fork and make sure that the build with enabled `zkcrypto` passes.

shreyas-londhe commented:
Thanks for pointing this out @sauliusgrigaitis. The latest commit on my branch passes the CI tests with both `bls-blst` and `bls-zkcrypto` features enabled [here](https://github.com/shreyas-londhe/grandine/blob/feat/modular-bls/.github/workflows/ci.yml#L40). Please let me know if you have any more suggestions. 

I am also planning to write a consistency check macro for testing the newly added backends with the `blst` backend so the issues I faced won't be repeated when more backends are added.

povi commented:
Looks good! There are some minor things though that I'd suggest to change, I'll write them down.

povi commented on bls/bls-core/src/error.rs:
```diff
@@ -0,0 +1,23 @@
+    #[error("invalid secret key")]
+    InvalidSecretKey,
+    #[error("invalid signature")]
+    InvalidSignature,
```
**[issue](https://conventionalcomments.org/)**: `InvalidSignature` error is not used anywhere. Please remove

shreyas-londhe replied:
@povi When we remove the `BLST_ERROR` from `bls-core`, `InvalidSignature` is used in `signature.rs`. This change is reflected in the latest commit - https://github.com/shreyas-londhe/grandine/blob/feat/modular-bls/bls/bls-blst/src/signature.rs#L42

povi commented on bls/bls-blst/src/signature.rs:
```diff
@@ -44,9 +43,12 @@ impl TryFrom<SignatureBytes> for Signature {
+
     #[must_use]
-    pub fn verify(self, message: impl AsRef<[u8]>, public_key: PublicKey) -> bool {
+    fn verify(&self, message: impl AsRef<[u8]>, public_key: &Self::PublicKey) -> bool {
```
**[nitpick](https://conventionalcomments.org/)**: `PublicKey` trait is `Copy`, it does not need to be passed as a reference. Pass `public_key` by value, as it was before.

shreyas-londhe replied:
@povi Thank you for the suggestion, but isn't it better if the object is passed by reference where it can be without used the need of copying. I would love to know the thinking behind this.

povi replied:
It depends on the size and the type of the data. Does it make sense to pass a large `struct` by reference? Absolutely. Does it make sense to pass `usize` by reference? Definitely not.

In this case, `PublicKey` is very small, fixed size `struct`. Performance-wise, pushing and popping few times to and from the stack is arguably better than accessing the heap by a reference. In addition to this, code is simpler and cleaner - no need to deal with references.

shreyas-londhe replied:
Thanks for such a detailed explanation, it makes sense. I will make the change.

sauliusgrigaitis commented on bls/bls-core/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+edition = { workspace = true }
+
+[dependencies]
+blst = { workspace = true }
```
We need to get rid of `blst` dependency here. Otherwise we still need to compile `blst` if we don't use `blst` as a backend.

sauliusgrigaitis commented on bls/Cargo.toml:
```diff
@@ -6,24 +6,11 @@ authors = ["Grandine <info@grandine.io>"]
-typenum = { workspace = true }
-zeroize = { workspace = true }
+[features]
+bls-blst = ["dep:bls-blst"]
```
Is there a particular reason why `bls` prefix is used for feature names? I prefer just `blst`, `zkcrypto`. As I mentioned previously we most likely will try to also hook `rust-kzg` backends using the same switches.

sauliusgrigaitis commented:
Did you find any way to use `blst` backend by default without explicitly specifying it always? It's very inconvenient.

shreyas-londhe commented:
> Did you find any way to use blst backend by default without explicitly specifying it always? It's very inconvenient.

@sauliusgrigaitis I'm not sure about this one. I tried uses `workspace.features` but the don't seem to link the features correctly. @povi Do you have any suggestions on this one?

povi commented:
> @povi Do you have any suggestions on this one?

No, sorry, I don't have

shreyas-londhe commented:
@sauliusgrigaitis I have one solution using `workspace.metadata.commands`, but I don't think that is suitable in this case.

ArtiomTr commented:
Hey @shreyas-londhe, I've faced the same issue in #88, and solved it by specifying `blst` as default feature in `grandine` crate. Try this patch:

```diff
From 10a946bb54ab25156bdbbc40a31d72933b098dfa Mon Sep 17 00:00:00 2001
From: sirse <artiom.tretjakovas2@gmail.com>
Date: Thu, 23 Jan 2025 19:36:41 +0200
Subject: [PATCH] Mark blst as default feature

---
 grandine/Cargo.toml | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/grandine/Cargo.toml b/grandine/Cargo.toml
index 65f9077..d7387f3 100644
--- a/grandine/Cargo.toml
+++ b/grandine/Cargo.toml
@@ -73,6 +73,8 @@ tempfile = { workspace = true }
 test-case = { workspace = true }
 
 [features]
+default = ["bls/blst"]
+
 logger-always-write-style = []
 
 # `preset-any` and `network-any` should not be passed to Cargo.
-- 
2.43.0
```

shreyas-londhe commented:
@ArtiomTr Does your patch enable us to switch to different backends at compile time once a default has set?
Something like here: https://github.com/shreyas-londhe/grandine/blob/feat/modular-bls/.github/workflows/ci.yml#L40

ArtiomTr commented:
Yes, you can do something like:
```
cargo test --release --no-fail-fast --no-default-features --features zkcrypto
```

shreyas-londhe commented:
@ArtiomTr Thanks for the suggestion. @sauliusgrigaitis @povi I have updated the CI to use the default `bls` backend `blst` and we can explicitly specify the `zkcrypto` backend by calling `--no-default-features --features bls/zkcrypto`. 

I hope this is satisfactory functionality. Please let me know if anything else is required for this PR to be merged.

povi commented on bls/src/secret_key.rs:
```diff
@@ -1,98 +0,0 @@
-
-use crate::{consts::DOMAIN_SEPARATION_TAG, Error, PublicKey, SecretKeyBytes, Signature};
-
-// `RawSecretKey` already implements `Zeroize` (with `zeroize(drop)`):
```
This pull request is ready to merge except for one last thing: `bls` crate lost most of the comments like this during the refactor. Could you put them back in appropriate places? Thanks

shreyas-londhe replied:
@povi Thanks for pointing this out. I've added the comments back into relevant places.

povi reviewed (approved):
Looks good. Thank you for your contribution!
